File: app/src/androidTest/java/br/com/codeteam/ctanywhere/PrefTest.java
package br.com.codeteam.ctanywhere;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.util.Log;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import br.com.codeteam.ctanywhere.preferences.Pref;

import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;
import static org.junit.Assert.assertEquals;

/**
 * Teste unitário para testar a classe de preferences
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class PrefTest {
    private static final String TAG = "PrefTest";

    private static final String NAME = "Madeleine";
    private static final String NAME_CHANGE = "Julieta";
    private static final String KEY_NAME = "NAME";

    private static final boolean TRUE = true;
    private static final String KEY_TRUE = "TRUE";

    private Context appContext;

    @Before
    public void setup() {
        this.appContext = InstrumentationRegistry.getTargetContext();
    }

    @Test
    public void write() {
        Pref.INSTANCE.write(KEY_NAME, NAME, this.appContext);
        assertEquals(Pref.INSTANCE.read(KEY_NAME, this.appContext), NAME);

        Pref.INSTANCE.write(KEY_TRUE, TRUE, this.appContext);
        assertTrue(Pref.INSTANCE.<Boolean>read(KEY_TRUE, this.appContext));
    }

    @Test
    public void clear() {
        this.write();

        Pref.INSTANCE.clearAll(this.appContext);
        assertNull(Pref.INSTANCE.read(KEY_NAME, this.appContext));
    }

    @Test
    public void registerListener() {
        this.write();

        Pref.INSTANCE.registerPreferenceChanged(this.appContext, new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
                Log.d(TAG, "onSharedPreferenceChanged: " + Pref.INSTANCE.read(key, appContext));
                assertEquals(Pref.INSTANCE.read(key, appContext), NAME_CHANGE);
            }
        });

        Pref.INSTANCE.write(KEY_NAME, NAME_CHANGE, this.appContext);
    }
}

File: app/src/androidTest/java/br/com/codeteam/ctanywhere/PrefTest.kt
package br.com.codeteam.ctanywhere

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.test.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import br.com.codeteam.ctanywhere.preferences.Pref
import br.com.codeteam.ctanywhere.preferences.PrefCrypto
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Teste unitário para testar a classe de preferences
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */
@RunWith(AndroidJUnit4::class)
class PrefTest {

    lateinit var context: Context

    @Before
    fun setup() {
        context = InstrumentationRegistry.getTargetContext()

        PrefCrypto.init(context)
    }

    @Test
    fun write() {
        Pref.write(KEY_NAME, NAME, context)
        assertEquals(Pref.read(KEY_NAME, context), NAME)

        Pref.write(KEY_TRUE, TRUE, context)
        assertTrue(Pref.read(KEY_TRUE, context))

        PrefCrypto.write(KEY_NAME, NAME)
        assertEquals(PrefCrypto.get(KEY_NAME), NAME)

        PrefCrypto.write(KEY_TRUE, TRUE)
        assertTrue(PrefCrypto.get(KEY_TRUE))
    }

    @Test
    fun clear() {
        this.write()

        Pref.clearAll(context)
        assertNull(Pref.read(KEY_NAME, context))

        PrefCrypto.deleteAll()
        assertNull(PrefCrypto.get(NAME))
    }

    @Test
    fun registerListener() {
        this.write()

        Pref.registerPreferenceChanged(context, SharedPreferences.OnSharedPreferenceChangeListener { sharedPreferences, key ->
            Log.d(TAG, "onSharedPreferenceChanged: " + Pref.read<Any>(key, context))
            assertEquals(Pref.read(key, context), NAME_CHANGE)
        })

        Pref.write(KEY_NAME, NAME_CHANGE, context)
    }

    @Test
    fun deletePrefCrypto() {
        PrefCrypto.write(KEY_NAME, NAME)
        PrefCrypto.delete(KEY_NAME)

        assertNull(PrefCrypto.get(KEY_NAME))
    }

    @Test
    fun countPrefCrypto() {
        PrefCrypto.deleteAll()
        PrefCrypto.write(KEY_NAME, NAME)
        assertEquals(PrefCrypto.count(), 1)

        PrefCrypto.deleteAll()
        assertEquals(PrefCrypto.count(), 0)
    }

    companion object {
        private val TAG = "PrefTest"

        private val NAME = "Madeleine"
        private val NAME_CHANGE = "Julieta"
        private val KEY_NAME = "NAME"

        private val TRUE = true
        private val KEY_TRUE = "TRUE"
    }
}
